### PR TITLE
ProgPOW Implementation

### PR DESCRIPTION
**Pull request by Liquid369**
## Issue being fixed or feature implemented
This changes the Mining Algorithm for SCC from X11 to Progressive POW used by Etherium. It is a GPU ***only*** Algorithm that will remove the use of ASIC's mining on SCC blockchain. There's been a lot of issues regarding block spacing, with pools and users jumping on and off, so with this, it also is a new difficulty algorithm that won't rise as exponentially. 

## What was done?
ProgPOW, New Difficulty Calculations

## How Has This Been Tested?
This was tested on testnet between a small group of users and SCC Team members. Upon finalizing testnet, we mimicked a few fork scenario's on mainnet to our liking.

## Breaking Changes
This is consensus breaking changes, because once the time and block heights kick in, the X11 hashing algorithm will be replaced with ProgPOW making older clients incompatible with the main chain.

## Checklist:
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by sc-9310 — APPROVED** [review]

**Comment by sc-9310** [comment]
[wip]
a few tests i've done: 
- sync mainnet from genesis to chaintip: successful
- compile fresh clone incl. deps: successful (ubuntu18, wsl2)

**Comment by sc-9310** [comment]
@JSKitty can you add some more checkpoints sometime cw42 please?

File: src/crypto/progpow/firopow_test_vectors.hpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+{
+    int block_number;
+    const char* header_hash_hex;
+    const char* boundary_hex;
+    const char* nonce_hex;
+    const char* mix_hash_hex;
+    const char* final_hash_hex;
+};   
+
+sccpow_hash_test_case sccopow_hash_test_cases[] = {
```
Comment (JSKitty) [inline src/crypto/progpow/firopow_test_vectors.hpp]: These tests don't currently apply to SCC (using Firo data), we could nuke 'em and populate them at a later date with our own hexes.

File: src/primitives/block.cpp [hunk 2]
```diff
@@ -4,28 +4,106 @@
...
+    if (IsProgPow()) {
+        return GetProgPowHashFull(mix_hash);
+    }
+    return GetHash();
+}
+
+uint256 CBlockHeader::GetPoWHash(int nHeight) const
 {
+    uint256 powHash, mix_hash;
+    /*TODO: FIX!*/
```
Comment (JSKitty) [inline src/primitives/block.cpp]: Is this an old comment or :eyes:

File: src/rpc/mining.cpp [hunk 3]
```diff
@@ -465,17 +496,20 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
...
-        throw JSONRPCError(RPC_CLIENT_IN_INITIAL_DOWNLOAD, PACKAGE_NAME " is in initial sync and waiting for blocks...");
+        if (::ChainstateActive().IsInitialBlockDownload())
+            throw JSONRPCError(RPC_CLIENT_IN_INITIAL_DOWNLOAD, PACKAGE_NAME " is in initial sync and waiting for blocks...");
 
-    // next bock is a superblock and we need governance info to correctly construct it
-    if (AreSuperblocksEnabled()
-        && !masternodeSync.IsSynced()
-        && CSuperblock::IsValidBlockHeight(::ChainActive().Height() + 1))
-            throw JSONRPCError(RPC_CLIENT_IN_INITIAL_DOWNLOAD, PACKAGE_NAME "is syncing with network...");
+        // next bock is a superblock and we need governance info to correctly construct it
```
Comment (JSKitty) [inline src/rpc/mining.cpp]: :chicken: bock --> block

**Review by JSKitty — APPROVED** [review]
The majority of the PR is (clearly) ProgPOW / ethash / keccak / scrypt changes :pick: these seem fine and standard to me, nothing in the rest of the client stands out bad to me either.

I left some *super trivial / nit* notes, but nothing major at all spotted. :rocket:

File: src/crypto/progpow/firopow_test_vectors.hpp (hunk 1 above)
Comment (Liquid369) [reply src/crypto/progpow/firopow_test_vectors.hpp]: Nuked.

File: src/primitives/block.cpp (hunk 2 above)
Comment (Liquid369) [reply src/primitives/block.cpp]: Yes it was thanks for pointing it out. Removed!

File: src/rpc/mining.cpp (hunk 3 above)
Comment (Liquid369) [reply src/rpc/mining.cpp]: Bock bock.

**Comment by Liquid369** [comment]
Pushed fixes

What I see left is just us making sure of checkpoints/minimumchainwork/defaultassumevalid

**Comment by Liquid369** [comment]
Update:
Compile hashes from latest commit
```d99acfefae701ba0fba1bb4a63789c3d5b3fe94da319deae3a6942c369ff426e```  scc-758057ad14fb-aarch64-linux-gnu-debug.tar.gz
```1d1677e68f3eb1c02ea498d457ef38be8b3965751f21736f4d613984639f864c```  scc-758057ad14fb-aarch64-linux-gnu.tar.gz
```fa0a36d7455bd62e8f58b92ce695196bd42608d80e988328b916803db5b04047```  scc-758057ad14fb-riscv64-linux-gnu-debug.tar.gz
```b4dbd843d446d6ace47b49d31cf1f72b46ad756c8449b2b584c825f00c4b74e3```  scc-758057ad14fb-riscv64-linux-gnu.tar.gz
```589b6c8956f8834e6a68485921abc48032b411adc5d9a27d9928ba241b2f4615```  scc-758057ad14fb-x86_64-linux-gnu-debug.tar.gz
```24792af81dc5016b9c3ed2627f291dd62bb87ae4d3d9f8fda2e24fbadc49b94d```  scc-758057ad14fb-x86_64-linux-gnu.tar.gz
```21e6af4aef42f89e48ddb50df2c3b663e8c8427b19b930cadd91ff8d5d29db28```  scc-758057ad14fb-win-unsigned.tar.gz
```50a32b673d02840801b70be1e2968c12b7292e5f69169c719c0d879a6f406395```  scc-758057ad14fb-win64-setup-unsigned.exe
```1ad9281617eab0e96503fe91658dcacb5da278fe71e499d95f68b7de836261cf```  scc-758057ad14fb-win64.zip
```109a6adc9ebcf9940913ebd8897809edf36b8b4b2f52bff9bfa9d446769b77ed```  scc-758057ad14fb-osx-unsigned.dmg
```1a6aa9db81bf10827acc97879111d70703e905e88a851d46660f32c3f68c4e4a```  scc-758057ad14fb-osx-unsigned.tar.gz
```1249cbaade58dd615ad9f4b48f5e4041bf0be73bb735963d53995ad24696aa8c```  scc-758057ad14fb-osx64-debug.tar.gz
```80d3ff588fe3526b4fb8725d4542716b9b802cb551a7325d60b1effec2dcf17e```  scc-758057ad14fb-osx64.tar.gz

**Pull request merged**